## Pipeline in Hugging Face

https://huggingface.co/docs/transformers/main_classes/pipelines#natural-language-processing


!pip install transformers

In [2]:
from transformers import pipeline

### 1. Text Summarization

- One can also choose from the other options of models that have been fine-tuned for the summarization task - bart-large-cnn, t5-small, t5-large, t5-3b, t5-11b. 
- You can check out the complete list of available models form https://huggingface.co/models?pipeline_tag=summarization&sort=downloads

In [3]:
summarizer = pipeline(
    "summarization", model="t5-base", tokenizer="t5-base", framework="tf"
)

config.json:   0%|          | 0.00/1.21k [00:00<?, ?B/s]

f:\Anaconda3\envs\torch_env\Lib\site-packages\huggingface_hub\file_download.py:143: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\user\.cache\huggingface\hub\models--t5-base. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)


Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`


model.safetensors:   0%|          | 0.00/892M [00:00<?, ?B/s]

All PyTorch model weights were used when initializing TFT5ForConditionalGeneration.

All the weights of TFT5ForConditionalGeneration were initialized from the PyTorch model.
If your task is similar to the task the model of the checkpoint was trained on, you can already use TFT5ForConditionalGeneration for predictions without further training.


spiece.model:   0%|          | 0.00/792k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.39M [00:00<?, ?B/s]

Device set to use 0


In [ ]:
input = """Parents need to know that Top Gun is a blockbuster 1980s action thriller starring Tom Cruise that's chock full of narrow escapes, 
chases, and battles. But there are also violent and upsetting scenes, particularly the death of a main character, which make it too intense for younger kids. 
There's also one graphic-for-its-time sex scene (though no explicit nudity) and quite a few shirtless men in locker rooms and, in one iconic sequence, 
on a beach volleyball court. Winning is the most important thing to all the pilots, who try to intimidate one another with plenty of posturing and banter 
-- though when push comes to shove, loyalty and friendship have important roles to play, too. While sexism is noticeable and almost all characters are men, 
two strong women help keep some of the objectification in check."""

"Parents need to know that Top Gun is a blockbuster 1980s action thriller starring Tom Cruise that's chock full of narrow escapes, \nchases, and battles. But there are also violent and upsetting scenes, particularly the death of a main character, which make it too intense for younger kids. \nThere's also one graphic-for-its-time sex scene (though no explicit nudity) and quite a few shirtless men in locker rooms and, in one iconic sequence, \non a beach volleyball court. Winning is the most important thing to all the pilots, who try to intimidate one another with plenty of posturing and banter \n-- though when push comes to shove, loyalty and friendship have important roles to play, too. While sexism is noticeable and almost all characters are men, \ntwo strong women help keep some of the objectification in check."

In [6]:
summarizer(input, min_length=5, max_length=30)

[{'summary_text': '1980s action thriller starring Tom Cruise is chock full of chases and battles . there are also violent and upsetting scenes,'}]

## 2. Question Answering

- full list of available models for the Question-Answering task.

  https://huggingface.co/models?pipeline_tag=question-answering

In [7]:
qa_pipeline = pipeline(model="deepset/roberta-base-squad2")

config.json:   0%|          | 0.00/571 [00:00<?, ?B/s]

f:\Anaconda3\envs\torch_env\Lib\site-packages\huggingface_hub\file_download.py:143: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\user\.cache\huggingface\hub\models--deepset--roberta-base-squad2. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)


model.safetensors:   0%|          | 0.00/496M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/79.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/899k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/772 [00:00<?, ?B/s]

Device set to use cuda:0


In [8]:
qa_pipeline(
    question="Where do I work?",
    context="I work as a Data Scientist at a lab in University of Montreal. I like to develop my own algorithms.",
)

{'score': 0.6422625184059143,
 'start': 39,
 'end': 61,
 'answer': 'University of Montreal'}

## 3. Named Entity Recognition

- 문장 속에서 특정한 의미를 가진 단어(=개체명)를 찾아내고, 그 종류(=엔티티의 유형)**를 분류하는 자연어처리(NLP) 기술

- NER 시스템은 보통 아래와 같은 유형으로 분류합니다:

    PERSON : 사람 이름

    ORGANIZATION : 조직, 회사, 기관

    LOCATION : 지명, 장소

    DATE : 날짜, 시간

    MONEY :	금액

    EVENT :	사건, 행사

 - The full list of models for text classification 

   https://huggingface.co/models?pipeline_tag=token-classification&sort=downloads

In [9]:
ner_classifier = pipeline(
    model="dslim/bert-base-NER-uncased", aggregation_strategy="simple"
)

config.json:   0%|          | 0.00/1.26k [00:00<?, ?B/s]

f:\Anaconda3\envs\torch_env\Lib\site-packages\huggingface_hub\file_download.py:143: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\user\.cache\huggingface\hub\models--dslim--bert-base-NER-uncased. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)


model.safetensors:   0%|          | 0.00/438M [00:00<?, ?B/s]

Some weights of the model checkpoint at dslim/bert-base-NER-uncased were not used when initializing BertForTokenClassification: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight']
- This IS expected if you are initializing BertForTokenClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForTokenClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


tokenizer_config.json:   0%|          | 0.00/39.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

Device set to use cuda:0


In [10]:
sentence = "I like to travel in Montreal."
entity = ner_classifier(sentence)
print(entity)

[{'entity_group': 'LOC', 'score': 0.9976744, 'word': 'montreal', 'start': 20, 'end': 28}]


'entity_group': 'LOC'	추출된 단어는 위치(Location), 즉 지명임을 의미합니다. LOC = location.

'score': 0.9976744	모델이 이 단어가 지명일 확률을 99.77%로 예측했다는 의미입니다. (자신감 점수)

'word': 'montreal'	추출된 개체명은 montreal입니다. 모델이 소문자로 예측했지만 실제 문장에서는 대문자 Montreal로 쓰였습니다.

'start': 20	문장에서 시작 인덱스입니다. 20번째 문자부터 시작하며, Python 인덱스 기준입니다.

'end': 28	문장에서 끝 인덱스입니다. 28번째 문자 전까지 (end는 exclusive) 포함합니다. 즉 "Montreal"은 sentence[20:28]로 추출됩니다.

## Text Classification 

- The full list of models for text classification 

  https://huggingface.co/models?pipeline_tag=text-classification

In [12]:
text_classifier = pipeline(
    model="distilbert-base-uncased-finetuned-sst-2-english"
)

Device set to use cuda:0


In [13]:
text_classifier("This movie is horrible!")

[{'label': 'NEGATIVE', 'score': 0.9997865557670593}]

In [14]:
text_classifier("I loved the narration of the movie!")

[{'label': 'POSITIVE', 'score': 0.9998612403869629}]

## Text Generation

- The full list of models for text classification 

  https://huggingface.co/models?pipeline_tag=text-generation

In [15]:
text_generator = pipeline(model="gpt2")

Device set to use cuda:0


In [18]:
prompt = "If it is sunny today then "

In [19]:
text_generator(prompt, do_sample=False)

The following generation flags are not valid and may be ignored: ['temperature']. Set `TRANSFORMERS_VERBOSITY=info` for more details.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


[{'generated_text': 'If it is sunny today then \xa0it will be cloudy tomorrow.\nI have been using this for a while now and I am very happy with it. I have been using it for a while now and I am very happy with it. I have been using it for a while now and I am very happy with it. I have been using it for a while now and I am very happy with it. I have been using it for a while now and I am very happy with it. I have been using it for a while now and I am very happy with it. I have been using it for a while now and I am very happy with it. I have been using it for a while now and I am very happy with it. I have been using it for a while now and I am very happy with it. I have been using it for a while now and I am very happy with it. I have been using it for a while now and I am very happy with it. I have been using it for a while now and I am very happy with it. I have been using it for a while now and I am very happy with it. I have been using it for a while now and I am very happy wit

In [20]:
# 텍스트 생성
output = text_generator(prompt, do_sample=False, max_new_tokens=50)

The following generation flags are not valid and may be ignored: ['temperature']. Set `TRANSFORMERS_VERBOSITY=info` for more details.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


In [21]:
generated_text = output[0]['generated_text']
print("Prompt:")
print(prompt)
print("\n Generated Continuation:")
print(generated_text[len(prompt):])  # 프롬프트 제외하고 생성된 부분만 출력

print("\n Full Generated Text:")
print(generated_text)  # 전체 출력 (프롬프트 + 생성된 문장)

Prompt:
If it is sunny today then 

 Generated Continuation:
 it will be cloudy tomorrow.
I have been using this for a while now and I am very happy with it. I have been using it for a while now and I am very happy with it. I have been using it for a while

 Full Generated Text:
If it is sunny today then  it will be cloudy tomorrow.
I have been using this for a while now and I am very happy with it. I have been using it for a while now and I am very happy with it. I have been using it for a while


### 다양한 설정 적용

- max_new_tokens	새로 생성할 최대 토큰 수 :50 (기본), 100, 200 등
- temperature	창의성 조절 (낮을수록 보수적, 높을수록 다양함) : 1.0 (기본), 0.7 (더 보수적), 1.5 (더 창의적)
- top_k	상위 k개 후보만 고려 (값이 작을수록 안정적)	: 50, 100, 0 (제한 없음)
- top_p	누적 확률이 p 이상인 후보만 고려 (nucleus sampling)	:  0.9, 0.8, 1.0
- do_sample	True면 샘플링 (랜덤), False면 greedy (가장 확률 높은 것 선택) :	True 또는 False

In [22]:
# 텍스트 생성
output = text_generator(
    prompt,
    max_new_tokens=50,
    temperature=0.9,   # 다양성 조절
    top_k=50,          # 상위 50개 후보 중에서 샘플링
    top_p=0.95,        # 누적 확률 95% 이상에서만 샘플링
    do_sample=True     # 랜덤 샘플링 사용
)

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


In [24]:
# 출력 정리
generated_text = output[0]['generated_text']
print("Prompt:")
print(prompt)
print("\n Generated Continuation:")
print(generated_text[len(prompt):])  # 생성된 부분만 출력

print("\n Full Generated Text:")
print(generated_text)

Prompt:
If it is sunny today then 

 Generated Continuation:
 let's try out a different color. It's much cooler in my case, but it does make some cool shadows.  I'm going to go with a lighter shade of pink and then some of the darker shadows, but this color works

 Full Generated Text:
If it is sunny today then  let's try out a different color. It's much cooler in my case, but it does make some cool shadows.  I'm going to go with a lighter shade of pink and then some of the darker shadows, but this color works


## Text Translation

- Other advanced models  

  https://huggingface.co/models?pipeline_tag=translation

In [26]:
# 번역 파이프라인 생성
en_fr_translator = pipeline("translation_en_to_fr", model="t5-small")

Device set to use cuda:0


In [27]:
# 입력 문장
text = "Hi, How are you?"

In [28]:
# 번역 수행
output = en_fr_translator(text, max_length=40, clean_up_tokenization_spaces=True)

Both `max_new_tokens` (=256) and `max_length`(=40) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


In [29]:
# 결과 출력
print(" Original (EN):")
print(text)

print("\n Translated (FR):")
print(output[0]['translation_text'])

print("\n Full Output Dictionary:")
print(output[0])

 Original (EN):
Hi, How are you?

 Translated (FR):
Bonjour, Comment êtes-vous?

 Full Output Dictionary:
{'translation_text': 'Bonjour, Comment êtes-vous?'}


### 한국어 → 영어 번역 (KO → EN)

In [46]:
from transformers import pipeline

# 한국어 → 영어 번역 파이프라인 생성
ko_en_translator = pipeline("translation", model="Helsinki-NLP/opus-mt-ko-en")

# 번역할 문장
text_ko = "저는 주말에 한국 영화를 보는 것을 좋아해요."

# 번역 수행
output = ko_en_translator(text_ko, max_length=100)

# 결과 출력
print("원문 (KO):", text_ko)
print("번역 (EN):", output[0]['translation_text'])

Device set to use cuda:0


원문 (KO): 저는 주말에 한국 영화를 보는 것을 좋아해요.
번역 (EN): I love watching Korean movies on weekends.
